sbi/analysis/sbc.py

```diff
@@ -0,0 +1,306 @@
+import warnings
+from typing import Callable, Dict, Tuple
+
+import torch
+from joblib import Parallel, delayed
+from torch import Tensor, ones, zeros
+from torch.distributions import Distribution, Uniform
+from tqdm.auto import tqdm
+
+from sbi.inference.posteriors.base_posterior import NeuralPosterior
+from sbi.simulators.simutils import simulate_in_batches, tqdm_joblib
+from sbi.utils.metrics import c2st
+
+
+def run_sbc(
+    prior: Distribution,
+    simulator: Callable,
+    posterior: NeuralPosterior,
+    num_sbc_samples: int = 1000,
+    num_posterior_samples: int = 100,
+    num_workers: int = 1,
+    sbc_batch_size: int = 1,
+    show_progress_bar: bool = True,
+) -> Tuple[Tensor, Tensor, Tensor]:
+    """Run simulation-based calibration (parallelized).
+
+    Returns sbc ranks, log probs of the true parameters under the posterior and samples
+    from the data averaged posterior, one for each sbc run, respectively.
+
+    Args:
+        prior: prior distribution.
+        simulator: simulator (or likelihood) that can called to generate observations.
+        posterior: a posterior obtained from sbi.
+        num_sbc_samples: number of sbc runs, i.e., number of repeated inferences.
+            Should be high, ~100 to give reliable results.
+        num_posterior_samples: number of approximate posterior samples used for ranking.
+        num_workers: number of CPU cores to use in parallel for running num_sbc_samples inferences.
+        sbc_batch_size: batch size for workers.
+        show_progress_bar: whether to display a progress over sbc runs.
+
+    Returns:
+        ranks: ranks of the ground truth parameters under the inferred posterior.
+        log_probs: log probs of the ground truth parameters under the inferred posterior.
+        dap_samples: samples from the data averaged posterior.
+    """
+
+    if num_sbc_samples < 1000:
+        warnings.warn(
+            """Number of SBC samples should be on the order ~100 to give realiable
+            results. We recommend using 1000."""
+        )
+    if num_posterior_samples < 100:
+        warnings.warn(
+            """Number of posterior samples for ranking should be on the order
+            of ~100 to give reliable SBC results. We recommend using at least 100."""
+        )
+
+    # Sample ground-truth parameters and simulator xos for SBC.
+    thos = prior.sample((num_sbc_samples,))
+    xos = simulate_in_batches(
+        simulator,
+        thos,
+        sim_batch_size=1000,
+        num_workers=num_workers,
+        show_progress_bars=show_progress_bar,
+    )
```

Comment:
I would strongly prefer that the user just passes `thetas` and `xs` instead of `prior` and `simulator`. This limitation would prevent me from running sbc on many of my models.